### ISSUE TITLE
Race in signal.c causes programs to hang

### ISSUE THREAD

**Issue by talex5** [ocaml/ocaml#12253]
When a signal is pending, `caml_process_pending_signals_exn` first clears the pending flag and then calls `caml_execute_signal_exn`, which masks the signal while the handler is running.

If another signal occurs after clearing the flag but before setting the mask then the handler runs with a pending signal. This causes `caml_enter_blocking_section` to go into an infinite loop.

To reproduce the bug easily, add a delay to `caml_execute_signal_exn`, just before masking the signal. After adding a `sleep(1)` there, this program hangs reliably on 5.0.0 and on the current 5.1 branch (5.1.0-alpha1-11-g48e0f9f59f):

```ocaml
let handle (_ : int) =
  print_endline "-> signal handler";
  Unix.sleepf 1.5;
  let rec aux () =
    match Unix.(waitpid [WNOHANG]) (-1) with
    | _, _ -> print_endline "reaped one child"; aux ()
    | exception Unix.Unix_error (Unix.ECHILD, _, _) -> ()
  in
  aux ();
  print_endline "<- signal handler"

let () =
  Sys.(set_signal sigchld) (Signal_handle handle);
  let _ = Unix.(create_process "/usr/bin/sleep" [| "sleep"; "0.1" |] stdin stdout stderr) in
  let _ = Unix.(create_process "/usr/bin/sleep" [| "sleep"; "0.2" |] stdin stdout stderr) in
  Unix.sleep 2
```

`strace` shows this continuously:

```
rt_sigprocmask(SIG_BLOCK, NULL, [CHLD], 8) = 0
rt_sigprocmask(SIG_BLOCK, NULL, [CHLD], 8) = 0
rt_sigprocmask(SIG_BLOCK, NULL, [CHLD], 8) = 0
rt_sigprocmask(SIG_BLOCK, NULL, [CHLD], 8) = 0
```

Originally reported against Eio at https://github.com/ocaml-multicore/eio/issues/495.

**Comment by gasche** [comment]
(I would think of @gadmm, @sadiqj or @Engil as people who could be interested in this report.)

**Comment by gadmm** [comment]
It sounds similar to a bug that has been fixed for a while at #11307, can you please try to reproduce with the branch over there?

**Comment by talex5** [comment]
> It sounds similar to a bug that has been fixed for a while at https://github.com/ocaml/ocaml/pull/11307, can you please try to reproduce with the branch over there?

I can't reproduce it with that branch.

**Comment by kayceesrk** [comment]
@gadmm given that the progress on #11307 has been slow, I wondered whether the fix to this bug is one way of unblocking progress there. Which of the commits in #11307 contribute to the fix for the current issue and can we carve it out into its own PR so that the bug gets fixed in 5.1 (possibly)? wdyt?

**Comment by gadmm** [comment]
Cross-reference: https://github.com/ocaml/ocaml/pull/11307#issuecomment-1597523166 for a proposed solution.

**Comment by gadmm** [comment]
@kayceesrk You have asked for a solution to this bug. I have proposed a few (see above). Do you have any feedback about the solutions, keeping the 5.1 target in mind?

**Comment by kayceesrk** [comment]
@gadmm Thanks for ping. Let's go with the patch that you've suggested so that it fixes the problem for 5.1. Would you have time to do this? I can quickly review and merge it. @smondet mentioned elsewhere that it is blocking 5.0 in production for them.

**Comment by kayceesrk** [comment]
@gadmm To get things moving, I've opened #12342 based on your patch. I've written down a detailed explanation of the issue and fix. Hopefully we can quickly merge this in time for 5.1.
